README.md

bittrex_database
----------------
is program for rebuilding and retrieving past nad current data on<br> 
all pairs from Bittrex.com.

- First we will install MariaDB and setup it.

Every following step is encapsulated in Docker:

- Main container (runs multiplied):
    - We'll create databases for 1, 5 and 60 minuts candlesticks</br>
      and create tables for each pair on Bittrex.com if not presented.</br>
      Each table has time as primary key and OHLC, Volume BTC Volume USD and Data Source.
    - We will refill history of 60 min. candlesticks if tables are completelly empty.
    - We start filling up databases with realtime data from Bittrex.com</br>
      using Tor circuit.

- Side containers for:
    - Backing up data.
    - Checking and repairing data consistency.

Tor circuit is used because multiple containers s planned to run.</br>
Bittrex.com allowes only limited frequncy for data request. Yet</br>
please be nice and use only max. three instances of data retrieval</br>
dockers.


changelog

				bittrex_database 0.1.0

Syntax for changelog:
	1.x.x - Marks major change of application. Previous APIs are no longer usable.
	x.1.x - Marks added functionality to application.
	x.x.1 - Marks minor changes and improvements of an application.

	1.0.0 - Marks first usable complete version of an applications.
	< 1.0.0 - Marks development of new application
	> 1.0.0 - Marks upgrades of an existing functional application

	Each 1.x.x and x.1.x changelog record is previously and equaly planned in BLUEPRINT
	and has equal tests in ./tests directory.



--- 0.2.0 - Project setup plus MariaDB script.
    Basic project and directory setup. Script for MariaDB installation and setup. Project planning.

--- 0.3.0 - Bittrex API
    Created Bittrex API. Some folder changes.

--- 0.4.0 - Databases creation
    Module for creating all neccessary databases.

--- 0.4.1 - Minor Databases.py changes
    Minor Databases.py tweaks.

--- 0.4.2 - Initial databases data fill and logging.
    Improved logging, structure and initial market and currencies info filling.



pytests/test_s1.py
#!/usr/bin/env python3
# Copyright (c) 2020 Jan Magrot

import time


def test_settings_reading():
	import lib.SettingsReader as Sr
	co = Sr.Config()
	pass


def test_logger():
	with open("./logs/app.log", "w") as o:
		o.write("")
	import logging
	import bittrex_database
	mlogger = logging.getLogger("bittrex_database." + __name__)
	mlogger.debug("Test.")
	with open("./logs/app.log", "r") as o:
		s = o.readline()
		s = s.split("\n")[0]
	assert "Test" in s
	assert "bittrex_database." + __name__ in s


def test_api():
	import lib.bittrex_api as Ba
	PublicApi = Ba.PublicApi()
	currencies = PublicApi.getAllCurrenciesInfo()
	time.sleep(1)
	assert isinstance(currencies, dict)
	assert len(currencies) > 200
	getAllMarketsInfo = PublicApi.getAllMarketsInfo()
	time.sleep(1)
	assert isinstance(getAllMarketsInfo, dict)
	assert len(getAllMarketsInfo) > 200
	getAllMarketsSummaries = PublicApi.getAllMarketsSummaries()
	time.sleep(1)
	assert isinstance(getAllMarketsSummaries, dict)
	assert len(getAllMarketsSummaries) > 200
	getAllMarketsTicker = PublicApi.getAllMarketsTicker()
	time.sleep(1)
	assert isinstance(getAllMarketsTicker, dict)
	assert len(getAllMarketsTicker) > 200
	s = PublicApi.pingBittrex()
	assert isinstance(s, dict)


lib/Databases.py
#!/usr/bin/env python3
# Copyright (c) 2021 janmagrot@gmail.com

import pymysql
import logging

module_logging = logging.getLogger("bittrex_database." + __name__)


def parse_sql_file(filename, variable=None, replacement=None):
	"""
	conn = pymysql.connect('test')
	stmts = parse_sql('my_sql_file.sql')
	with conn.cursor() as cursor:
        for stmt in stmts:
            cursor.execute(stmt)
    conn.commit()
	:param replacement:
	:param variable:
	:param filename:
	:return:
	"""
	data = open(filename, 'r').readlines()
	stmts = []
	DELIMITER = ';'
	stmt = ''

	for lineno, line in enumerate(data):
		if variable and replacement:
			if variable in line:
				line.replace(variable, replacement)

		if not line.strip():
			continue

		if line.startswith('--'):
			continue

		if 'DELIMITER' in line:
			DELIMITER = line.split()[1]
			continue

		if DELIMITER not in line:
			stmt += line.replace(DELIMITER, ';')
			continue

		if stmt:
			stmt += line
			stmts.append(stmt.strip())
			stmt = ''
		else:
			stmts.append(line.strip())
	return stmts


def createDatabases():
	"""
	Function for creating databases.
	:return:
	"""
	conn = pymysql.connect(host='localhost',
	                       user='jan',
	                       password='17051982')
	func_logging = logging.getLogger("bittrex_database." + str(__name__) + ".createDatabases()")
	stmts = parse_sql_file("./databases/create_databases.sql")
	with conn.cursor() as cursor:
		for stmt in stmts:
			cursor.execute(stmt)
	conn.commit()
	func_logging.info("Initial empty databases without columns, have been created.")
	return True


def createCurrenciesInfoDatabases(currencies, database):
	"""
	Function that creates tables for informations about
	currencies.
	:param database:
	:param currencies:
	:return:
	"""
	conn = pymysql.connect(host='localhost',
	                       user='jan',
	                       password='17051982',
	                       database=database)
	func_logging = logging.getLogger("bittrex_database." + str(__name__) + ".createCurrenciesInfoDatabases()")
	for ix in currencies:
		with conn.cursor() as cur:
			comm = f'CREATE OR REPLACE TABLE `{ix}` (' \
			       f'symbol CHAR(12) NOT NULL, ' \
			       f'name CHAR(20) NOT NULL, ' \
			       f'status CHAR(10) NOT NULL,' \
			       f'minConfirmations TINYINT NOT NULL ,' \
			       f'txFee DOUBLE,' \
			       f'baseAddress CHAR(180),' \
			       f'PRIMARY KEY (symbol)' \
			       f') ENGINE=Maria'
			cur.execute(comm)
	conn.commit()
	func_logging.info("Currencies info database have been initiated with empty columns.")
	return True


def createMarketsInfoDatabases(markets, database):
	"""
	Function that creates tables for informations about
	markets.
	:param database:
	:param markets:
	:return:
	"""
	conn = pymysql.connect(host='localhost',
	                       user='jan',
	                       password='17051982',
	                       database=database)
	func_logging = logging.getLogger("bittrex_database." + str(__name__) + ".createMarketsInfoDatabases()")
	for ix in markets:
		ix = "_".join(str(ix).split("-"))
		with conn.cursor() as cur:
			comm = f'CREATE OR REPLACE TABLE `{ix}` (' \
			       f'symbol CHAR(20) NOT NULL, ' \
			       f'baseCurrencySymbol CHAR(12) NOT NULL, ' \
			       f'quoteCurrencySymbol CHAR(12) NOT NULL, ' \
			       f'minTradeSize FLOAT NOT NULL, ' \
			       f'pprecision TINYINT NOT NULL, ' \
			       f'status CHAR(10) NOT NULL, ' \
			       f'PRIMARY KEY (symbol)' \
			       f') ENGINE=Maria'
			cur.execute(comm)
	conn.commit()
	func_logging.info("Markets info database have been initiated with empty columns.")
	return True


def createOneMinuteCandles(markets, database):
	"""
	Function that creates tables for one minute candles.
	:param database:
	:param markets:
	:return:
	"""
	conn = pymysql.connect(host='localhost',
	                       user='jan',
	                       password='17051982',
	                       database=database)
	func_logging = logging.getLogger("bittrex_database." + str(__name__) + ".createOneMinuteCandles()")
	for ix in markets:
		with conn.cursor() as cur:
			comm = f'CREATE OR REPLACE TABLE `{ix}` (' \
			       f'symbol CHAR(20) NOT NULL, ' \
			       f'ttime FLOAT NOT NULL, ' \
			       f'oopen DOUBLE NOT NULL, ' \
			       f'hhigh DOUBLE NOT NULL, ' \
			       f'llow DOUBLE NOT NULL, ' \
			       f'cclose DOUBLE NOT NULL, ' \
			       f'base_vol DOUBLE NOT NULL, ' \
			       f'quote_vol DOUBLE NOT NULL, ' \
			       f'usd_vol DOUBLE NOT NULL, ' \
			       f'PRIMARY KEY (symbol, ttime)' \
			       f') ENGINE=Maria'
			cur.execute(comm)
	conn.commit()
	func_logging.info("One minute candles database have been initiated with empty columns.")
	return True


def createFiveMinuteCandles(markets, database):
	"""
	Function that creates tables for five minute candles.
	:param database:
	:param markets:
	:return:
	"""
	conn = pymysql.connect(host='localhost',
	                       user='jan',
	                       password='17051982',
	                       database=database)
	func_logging = logging.getLogger("bittrex_database." + str(__name__) + ".createFiveMinuteCandles()")
	for ix in markets:
		with conn.cursor() as cur:
			comm = f'CREATE OR REPLACE TABLE `{ix}` (' \
			       f'symbol CHAR(20) NOT NULL, ' \
			       f'ttime FLOAT NOT NULL, ' \
			       f'oopen DOUBLE NOT NULL, ' \
			       f'hhigh DOUBLE NOT NULL, ' \
			       f'llow DOUBLE NOT NULL, ' \
			       f'cclose DOUBLE NOT NULL, ' \
			       f'base_vol DOUBLE NOT NULL, ' \
			       f'quote_vol DOUBLE NOT NULL, ' \
			       f'usd_vol DOUBLE NOT NULL, ' \
			       f'PRIMARY KEY (symbol, ttime)' \
			       f') ENGINE=Maria'
			cur.execute(comm)
	conn.commit()
	func_logging.info("Five minutes candles database have been initiated with empty columns.")
	return True


def createOneHourCandles(markets, database):
	"""
	Function that creates tables for one minute candles.
	:param database:
	:param markets:
	:return:
	"""
	conn = pymysql.connect(host='localhost',
	                       user='jan',
	                       password='17051982',
	                       database=database)
	func_logging = logging.getLogger("bittrex_database." + str(__name__) + ".createOneHourCandles()")
	for ix in markets:
		with conn.cursor() as cur:
			comm = f'CREATE OR REPLACE TABLE `{ix}` (' \
			       f'symbol CHAR(20) NOT NULL, ' \
			       f'ttime FLOAT NOT NULL, ' \
			       f'oopen DOUBLE NOT NULL, ' \
			       f'hhigh DOUBLE NOT NULL, ' \
			       f'llow DOUBLE NOT NULL, ' \
			       f'cclose DOUBLE NOT NULL, ' \
			       f'base_vol DOUBLE NOT NULL, ' \
			       f'quote_vol DOUBLE NOT NULL, ' \
			       f'usd_vol DOUBLE NOT NULL, ' \
			       f'PRIMARY KEY (symbol, ttime)' \
			       f') ENGINE=Maria'
			cur.execute(comm)
	conn.commit()
	func_logging.info("One hour candles database have been initiated with empty columns.")
	return True


def createAllDatabases(currencies, markets):
	"""
	Function for creating all the databases.
	:return:
	"""
	createDatabases()
	createCurrenciesInfoDatabases(currencies, "currenciesinfo")
	createMarketsInfoDatabases(markets, "marketinfo")
	createOneMinuteCandles(markets, "oneminutecandles")
	createFiveMinuteCandles(markets, "fiveminutescandles")
	createOneHourCandles(markets, "onehourcandles")
	module_logging.info("All databases have been created.")
	return True


def removeAllDatabases():
	conn = pymysql.connect(host='localhost',
	                       user='jan',
	                       password='17051982')
	func_logging = logging.getLogger("bittrex_database." + str(__name__) + ".removeAllDatabases()")
	stmts = parse_sql_file("./databases/delete_databases.sql")
	with conn.cursor() as cursor:
		for stmt in stmts:
			cursor.execute(stmt)
	conn.commit()
	func_logging.info("All created databases have been deleted.")
	return True


def insertCurrenciesInfo(currency, currinfo):
	"""
	Inserts informations about currencies
	to table.
	:return:
	"""
	func_logging = logging.getLogger("bittrex_database." + str(__name__) + ".insertCurrenciesInfo()")
	conn = pymysql.connect(host='localhost',
	                       user='jan',
	                       password='17051982',
	                       database="currenciesinfo")
	symbol = str(currinfo["symbol"])
	name = str(currinfo["name"])
	status = str(currinfo["status"])
	if "minConfirmations" in currinfo.keys():
		minConfirmations = int(currinfo["minConfirmations"])
	else:
		minConfirmations = ""
	if "txFee" in currinfo.keys():
		txFee = float(currinfo["txFee"])
	else:
		txFee = ""
	if "baseAddress" in currinfo.keys():
		baseAddress = str(currinfo["baseAddress"])
	else:
		baseAddress = ""
	try:
		with conn.cursor() as curr:
			comm = f"INSERT IGNORE INTO {currency}(symbol, name, status, minConfirmations, " \
			       f"txFee, baseAddress) VALUES('{symbol}','{name}','{status}',{minConfirmations}," \
			       f"{txFee},'{baseAddress}');"
			curr.execute(comm)
		conn.commit()
	except Exception:
		func_logging.exception("Exception during inserting currencies informations.")
	func_logging.debug("Databases about currencies info have been filled with data.")
	return True


def insertMarketsInfo(market, marketinfo):
	"""
	Inserts informations about currencies
	to table.
	:return:
	"""
	func_logging = logging.getLogger("bittrex_database." + str(__name__) + ".insertMarketsInfo()")
	conn = pymysql.connect(host='localhost',
	                       user='jan',
	                       password='17051982',
	                       database="marketinfo")
	market = "_".join(str(market).split("-"))
	symbol = "_".join(str(marketinfo["symbol"]).split("-"))
	baseCurrencySymbol = str(marketinfo["baseCurrencySymbol"])
	quoteCurrencySymbol = str(marketinfo["quoteCurrencySymbol"])
	minTradeSize = float(marketinfo["minTradeSize"])
	pprecision = int(marketinfo["precision"])
	status = str(marketinfo["status"])
	try:
		with conn.cursor() as curr:
			comm = f"INSERT IGNORE INTO {market}(symbol, baseCurrencySymbol, quoteCurrencySymbol, minTradeSize, " \
			       f"pprecision, status) VALUES('{symbol}','{baseCurrencySymbol}','{quoteCurrencySymbol}',{minTradeSize}," \
			       f"{pprecision},'{status}');"
			curr.execute(comm)
		conn.commit()
	except Exception:
		func_logging.exception("Exception during inserting markets informations.")
	func_logging.debug("Databases about markets info have been filled with data.")
	return True


lib/Bittrex_api.py
#!/usr/bin/env python3
# Copyright (c) 2021 janmagrot@gmail.com

import requests
import logging
import requests.exceptions

"""
Module for simple data retrieving from Bittrex.com
"""

module_logging = logging.getLogger("bittrex_database." + __name__)
BASEURL = "https://api.bittrex.com/v3/"
false = bool(False)
true = bool(True)
null = None


class PublicApi:
	"""->
	Class handling requests and returns from Bittrex.com.
	"""

	def __init__(self):
		self._logging = logging.getLogger("bittrex_database." + str(__name__) + ".PublicApi()")
		self._logging.info("PublicApi class initiated.")

	@staticmethod
	def __getRaw(method):
		"""
		Method that takes BASEURL + method a send request to Bittrex.com
		servers. Used for encapsulation in excpetions method.
		"""
		request_url = BASEURL + method
		raw = requests.get(request_url, timeout=1.0)
		raw.encoding = 'utf-8'
		return raw

	def __getPublicQueryWrapper(self, method):
		"""
		Queries Bittrex with given method and options.
		:param method:
		:return:
		"""
		sc = None
		raw = None
		try:
			raw = self.__getRaw(method)
			sc = raw.status_code
			if sc != 200:
				raw.raise_for_status()
			else:
				response = raw.json()
				if raw.ok:
					self._logging.debug("Response is presented and succesfull.")
					return response
				else:
					msg = str(response["message"])
					self._logging.error("Connection problems. Response['success'] is negative. Return code is: {}, "
					                    ", meassage from server: {}, Whole response: {}".
					                    format(str(sc), str(msg),
					                           str(response)))
					return False
		except requests.HTTPError:
			self._logging.exception("Connection problem. Status Status code is: {},".format(str(sc)))
			return False
		except requests.ConnectionError:
			self._logging.exception("Connection problem. Status Status code is: {},".format(str(sc)))
			return False
		except ValueError as e:
			self._logging.exception("Problems with decoding data from server. "
			                        "Raw data {}. Details: {}".
			                        format(str(raw.text), str(e.args)))
			return False
		except requests.ReadTimeout:
			self._logging.exception("ReadTimeout.")
			return False
		except requests.Timeout:
			self._logging.exception("Timeout.")
			return False
		except BaseException:
			self._logging.exception("Exception occured")
			return False

	def getAllCurrenciesInfo(self):
		"""
		Get informations about all currencies on Bittrex market.
		"""
		s = self.__getPublicQueryWrapper("currencies")
		return {i["symbol"]: i for i in s}

	def getAllMarketsInfo(self):
		"""
		Get informations about all Bittrex Markets.
		:return:
		"""
		d = self.__getPublicQueryWrapper("markets")
		return {i["symbol"]: i for i in d}

	def getAllMarketsSummaries(self):
		"""
		Return informations about all markets simmaries within
		last 24 hours.
		:return:
		"""
		x = self.__getPublicQueryWrapper("markets/summaries")
		return {i["symbol"]: i for i in x}

	def getAllMarketsTicker(self):
		"""
		Returns ticker for all markets on Bittrex.
		:return:
		"""
		r = self.__getPublicQueryWrapper("markets/tickers")
		return {i["symbol"]: i for i in r}

	def pingBittrex(self):
		"""
		Verification Bittrex.com is online.
		:return:
		"""
		return dict(eval(self.__getRaw("ping").text))


Bittrex_database.py
#!/usr/bin/env python3
# Copyright (c) 2021 Jan Magrot

import logging
import lib.Bittrex_api as Ba
import lib.Databases as Db

logger = None


def Logger():
	# clean
	with open("./logs/app.log", "w") as o:
		o.write("")
	# Logger:
	global logger
	logger = logging.getLogger('bittrex_database')
	logger.setLevel(logging.INFO)
	fh = logging.FileHandler('./logs/app.log')
	fh.setLevel(logging.INFO)
	formatter = logging.Formatter('%(asctime)s - %(name)s - %(levelname)s - %(message)s')
	fh.setFormatter(formatter)
	logger.addHandler(fh)


Logger()


def BuildTables():
	Db.removeAllDatabases()
	pa = Ba.PublicApi()
	currinfo = pa.getAllCurrenciesInfo()
	marketsinfo = pa.getAllMarketsInfo()
	currencies = {k for k in currinfo.keys()}
	markets = {k for k in marketsinfo.keys()}
	Db.createAllDatabases(currencies, markets)
	#
	for k, v in currinfo.items():
		Db.insertCurrenciesInfo(k, v)
	for k, v in marketsinfo.items():
		Db.insertMarketsInfo(k, v)


BuildTables()


.devbin/bigcommit.sh
#!/bin/bash
# This script is supposed to run from Makefile in parent directory.
# Commit requires long and short description as well as version of commit number.
# TODO: zahrnout pytest testy a pylint. Zkusit prospector, mypy, Pylama.
if [ ! -d .git ]
then
	cd ..
	if [ ! -d .git ]
	then
		echo "
		I cant find .git directory! Terminating.
		"
		exit 1
	fi
fi

clear
echo "

Printing GIT STATUS:

"
sleep 1
git status

github () {
if [ -e .github ]
then
	git push origin master && echo -e "\nPushed to github.\n"
fi
}

commit () {
	# At first tests must pass.
	clear
	echo -e "\n\nTestuji:\n"
	SHFILES=$(find . -name "*.sh")
	for i in $SHFILES
	do
		chmod 755 "$i"
	done
	# Check for syntax errors, stdout to /dev/null:
	(
	for i in $SHFILES
	do
		bash -xn "$i"
	done
	) > /dev/null
	# Finegrain check with spellcheck:
	(
	for i in $SHFILES
	do
		shellcheck "$i"
	done
	)
	# Pytests:
	PYFILES=$(find . -name "*.py")
	python3 -m pytest pytests/
	#
	while true
	do
		echo -e "\nTesty provedeny. Chcete pokračovat s commitem? (y/n)"
		read -r anw
		case $anw in
		y|Y)
			break
			;;
		n|N)
			exit 0
			;;
		*)
			echo -e "\nNeplatný znak. Znovu.\n"
			sleeps 1s
			clear
			;;
		esac
	done
	echo "Enter changelog number: "
	read -r number
	echo "Enter short description: "
	read -r sdesc
	echo "Enter long description:
	"
	read -r ldesc
	head="$number - $sdesc"
	# Commit and push.
	echo "--- $head" >> changelog && echo "    $ldesc" >> changelog && echo "" >> changelog && git add changelog
	git commit -m "$head" -m "$ldesc" && github && \
	echo "
	Commit, kontrola a zápis do changelogu provedeny.
	" || echo "
	Commit a kontrola se nezdařily."
}

sleep 1
while true
do
	echo "

	Do You want to add some files and changes to git, before proceeding?
	Press 'c' for proceeding commit or 't' for terminating and manualy adding changes to git.
	"
	read -r answer

	case $answer in
	c|C)
		commit
		break
		;;
	t|T)
		break
		;;
	*)
		echo "
		Neplatný znak! Znovu.
		"
		sleep 1
		clear
		;;
	esac
done

exit 0


bash_scripts/init_mariadb.sh
#!/bin/bash

echo -e "

			INITIAL DATABASES:

This will take you through a series of prompts where you can make some changes to
your MariaDB installation’s security options.


1.	The first prompt will ask you to enter the current database root password. Since
	you have not set one up yet, press ENTER to indicate 'none'.

2.	The next prompt asks you whether you’d like to set up a database root password.
	On Ubuntu, the root account for MariaDB is tied closely to automated system
	maintenance, so we should not change the configured authentication methods for
	that account. Doing so would make it possible for a package update to break the
	database system by removing access to the administrative account. Type N and then
	press ENTER.

3.	From there, you can press Y and then ENTER to accept the defaults for all the
	subsequent questions. This will remove some anonymous users and the test
	database, disable remote root logins, and load these new rules so that MariaDB
	immediately implements the changes you have made.

ENTER TO CONTINUE....
"
read -r > /dev/null
sudo mysql_secure_installation
clear
echo -e "

			NOW DO THIS:

To this end, we will create a new account called admin with the same capabilities
as the root account, but configured for password authentication. Open up the
MariaDB prompt from your terminal:

	sudo mariadb

Then create a new user with root privileges and password-based access. Be sure to
change the username and password to match your preferences.

Write to shell:

	GRANT ALL ON *.* TO 'admin'@'localhost' IDENTIFIED BY 'password' WITH GRANT OPTION;

Flush the privileges to ensure that they are saved and available in the current
session:

	FLUSH PRIVILEGES;

Following this, exit the MariaDB shell:

	exit


ENTER TO CONTINUE....
"
read -r > /dev/null
sudo mariadb
clear
echo -e "
	PRINTING STATUS:
"
sudo systemctl start mariadb
sudo systemctl status mariadb
sudo mysqladmin version
echo -e "

ENTER TO CONTINUE....

"
read -r > /dev/null

clear
echo -e "

	Now checking user access---

ENTER:

mariadb -u user -ppassword

"

exit 0


lib/bittrex_api.py
#!/usr/bin/env python3
# Copyright (c) 2021 janmagrot@gmail.com

import requests
import logging
import requests.exceptions

"""
Module for simple data retrieving from Bittrex.com
"""

module_logging = logging.getLogger("bittrex_database." + __name__)
BASEURL = "https://api.bittrex.com/v3/"
false = bool(False)
true = bool(True)
null = None


class PublicApi:
	"""->
	Class handling requests and returns from Bittrex.com.
	"""

	def __init__(self):
		self._logging = logging.getLogger("bittrex_database." + str(__name__) + ".PublicApi()")
		self._logging.info("PublicApi class initiated.")

	@staticmethod
	def __getRaw(method):
		"""
		Method that takes BASEURL + method a send request to Bittrex.com
		servers. Used for encapsulation in excpetions method.
		"""
		request_url = BASEURL + method
		raw = requests.get(request_url, timeout=1.0)
		raw.encoding = 'utf-8'
		return raw

	def __getPublicQueryWrapper(self, method):
		"""
		Queries Bittrex with given method and options.
		:param method:
		:return:
		"""
		sc = None
		raw = None
		try:
			raw = self.__getRaw(method)
			sc = raw.status_code
			if sc != 200:
				raw.raise_for_status()
			else:
				response = raw.json()
				if raw.ok:
					self._logging.info("Response is presented and succesfull.")
					return response
				else:
					msg = str(response["message"])
					self._logging.error("Connection problems. Response['success'] is negative. Return code is: {}, "
					                    ", meassage from server: {}, Whole response: {}".
					                    format(str(sc), str(msg),
					                           str(response)))
					return False
		except requests.HTTPError:
			self._logging.exception("Connection problem. Status Status code is: {},".format(str(sc)))
			return False
		except requests.ConnectionError:
			self._logging.exception("Connection problem. Status Status code is: {},".format(str(sc)))
			return False
		except ValueError as e:
			self._logging.exception("Problems with decoding data from server. "
			                        "Raw data {}. Details: {}".
			                        format(str(raw.text), str(e.args)))
			return False
		except requests.ReadTimeout:
			self._logging.exception("ReadTimeout.")
			return False
		except requests.Timeout:
			self._logging.exception("Timeout.")
			return False
		except BaseException:
			self._logging.exception("Exception occured")
			return False

	def getAllCurrenciesInfo(self):
		"""
		Get informations about all currencies on Bittrex market.
		"""
		s = self.__getPublicQueryWrapper("currencies")
		return {i["symbol"]: i for i in s}

	def getAllMarketsInfo(self):
		"""
		Get informations about all Bittrex Markets.
		:return:
		"""
		s = self.__getPublicQueryWrapper("markets")
		return {i["symbol"]: i for i in s}

	def getAllMarketsSummaries(self):
		"""
		Return informations about all markets simmaries within
		last 24 hours.
		:return:
		"""
		s = self.__getPublicQueryWrapper("markets/summaries")
		return {i["symbol"]: i for i in s}

	def getAllMarketsTicker(self):
		"""
		Returns ticker for all markets on Bittrex.
		:return:
		"""
		s = self.__getPublicQueryWrapper("markets/tickers")
		return {i["symbol"]: i for i in s}

	def pingBittrex(self):
		"""
		Verification Bittrex.com is online.
		:return:
		"""
		return dict(eval(self.__getRaw("ping").text))


import json

Pa = PublicApi()
ci = Pa.getAllMarketsInfo()
#s = list(ci.keys())
s = json.dumps(ci, indent=4)
print(s)
